src/net/db.rs

```diff
@@ -1073,6 +1283,136 @@ impl PeerDB {
                     WHERE network_id = ?11 AND addrbytes = ?12 AND port = ?13", args)
             .map_err(db_error::SqliteError)?;
 
+        if let Some(old_peer) = old_peer_opt {
+            let slot_opt = Self::find_peer_slot(
+                tx,
+                neighbor.addr.network_id,
+                &neighbor.addr.addrbytes,
+                neighbor.addr.port,
+            )?;
+            if old_peer.public_key.to_bytes_compressed()
+                != neighbor.public_key.to_bytes_compressed()
+            {
+                // this peer has re-keyed, so it might be a new peer altogether.
+                // require it to re-announce its DBs
+                if let Some(slot) = slot_opt {
+                    debug!("Peer at slot {} changed; dropping its DBs", slot);
+                    PeerDB::drop_stacker_dbs(tx, slot)?;
+                }
+            }
+        }
+        Ok(())
+    }
+
+    /// Find a peer's slot in the DB.
+    /// The peer can have at most one slot.
+    /// Return Some(slot id) if the peer is inserted
+    /// Return None if not.
+    fn find_peer_slot(
+        conn: &Connection,
+        network_id: u32,
+        addrbytes: &PeerAddress,
+        port: u16,
+    ) -> Result<Option<u32>, db_error> {
+        // the peer can be at any of these slots, if it is present at all.
+        let slots = PeerDB::peer_slots(conn, network_id, addrbytes, port)?;
+        for slot in slots.into_iter() {
+            let used_slot = PeerDB::has_specific_peer_at(conn, network_id, addrbytes, port, slot)?;
+            if used_slot {
+                return Ok(Some(slot));
+            }
+        }
```

Comment:
Why does this function need to iterate? Can't the occupied slot, if the peer is present, be queried directly?


```
"SELECT slot FROM frontier WHERE network_id = ?1 AND addrbytes = ?2 AND port = ?3";
```

Reply:
Yup, will fix